develop/tutorials/articles/110-portlets/15-javascript/04-npm/08-npm-sdk.markdown

```diff
@@ -114,16 +113,19 @@ Now that you are a little more familiar with feature levels, you can dive deeper
 and learn how semantic versioning affects them.
 
 ### Semantic Versioning Compatibility [](id=semantic-versioning-compatibility)
-
-As long your project has components with the minimum version numbers, you can 
-use the features listed in the feature level (plus all those inherited from the 
-previous feature level, of course). If you set a component to a higher version 
-number that is compatible according to semantic versioning, you are still in 
-the same feature level (or maybe a higher one if you increase version numbers 
-enough to reach a new level). This can lead to running an unsupported npm SDK 
-version that may be broken or produce unexpected results. For example, say that 
-feature level 2 is the maximum possible, and your component versions are 
-slightly higher than the minimum versions required (Loader modules extender has 
+<!--This whole section is pretty vague. Maybe someone with a better knowledge
+base than me would understand it-NR--> As long your project has components with
+the minimum version numbers, you can use the features listed in the feature
+level (plus all those inherited from the previous feature level). If you set
+a component to a higher version number that is compatible according to semantic
+versioning, you are still in the same feature level (or maybe a higher one if
+you increase version numbers enough to reach a new level).<!--We need some
+precision here. If we have no idea how high is too high, shouldn't we be
+telling users to stick with the minimum version numbers? Is there any reason
+not to do this?-NR--> This can lead to running an unsupported npm SDK version
```

Comment:
This would be determined by the minimum component versions listed in the feature level. I agree the way it is phrased is confusing. I think you're right, we should just recommend that they stick to the minimum feature level listed and just say that "The component versions listed for the feature level are considered stable for the feature set. Setting a component to a higher version number than the minimum listed for the feature level may produce unexpected results."

Reply:
So there's no use case that might involve exceeding the minimum requirement?

Reply:
Well, in the example Ivan gave (he wrote the original README that this article was created from) you could theoretically have a component that is a higher version than the minimum listed for a feature level. His main point was that, that should be fine as long as it is still semantically compatible. Since components can have varying versions (they are not all updated at the same time), this is definitely a possible use case. The minimum versions are the ones that are considered tested, stable, and compatible (supported). So, the warning that follows about using higher versions is needed. Does that make sense?

Reply:
I understand it, but I'm a little uncomfortable recommending that users *not* use the latest versions. I mean, I assume Ivan wouldn't have bothered to explain that something *could* be done if there was no reason that it ever *should* be done.

Reply:
True. I think that it can be done, it's just that we can't verify that those versions will be compatible or stable. Because every component is different, I don't think there is a hard and fast rule for what versions will be compatible or stable for a given feature. The loose rule is that semantic versions determine compatibility. They are working on support for this as we speak. I'll try to clarify this with Ivan a bit more.

Reply:
I think that it can be done, it's just that we can't verify that those versions will be compatible or stable. Because every component is different, I don't think there is a hard and fast rule for what versions will be compatible or stable for a given feature. The loose rule is that semantic versions determine compatibility. They are working on support for this as we speak. I'll try to clarify this with Ivan a bit more.

Reply:
Thanks.

Reply:
So here is what Ivan had to say:
```
"we should recommend using the versions specified by a FL unless you use a newer patch (minor
number increase) version for a component. Of course, this would be a rule of thumb, that may 
not be followed under some circumstances."
```
So with npm, the component version you use is specified in your project's `package.json`. Node then downloads that version of the component for your project to use. Generally, I think the community is use to using specific versions of components in their projects to get certain features to work. This is the expected workflow. Since we have no control over what changes are made to the other components (`npm`, `Gradle`, and `Babel` are developed and owned by someone else) we can only support what our components do. So, I think recommending that they use the version specified for the feature level is the best approach, unless they're using a minor version patch as Ivan said. Using anything else may produce unstable or unexpected results.

Keep in mind that at this time there are only 2 feature levels listed, but more will be added, which will use later versions of the components. This is more of a tool that lets you verify what components and versions are required to use certain features in Liferay.

Reply:
Ok. I think I should note that there may be some specific cases when a user will want version numbers above the minimum required. If we have any docs on when or why someone would want to do that, we could link to them.

Otherwise, I think I should cut out any reference to a "minor number increase." Unless we can identify exactly what is a minor number and what isn't, I don't think we have any business talking about it.

How does that sound?

Reply:
I don't believe we have any docs on why someone would want to use a later version. That's up to the developer and is dependent on their individual project. Minor number refers to the semantic version number (Major.Minor.Patch). For example, 2.1.5 (2 is the major version, 1 is the minor number, and 5 is the patch.

Reply:
Oooooh.

Reply:
Ok. I understand. I'll go over those paragraphs again.

Reply:
Ok. I wielded Occam's Razor pretty vigorously this time, if you want to look over it and tell me what you think.